clar2wasm/src/words/equal.rs

```diff
@@ -715,3 +889,57 @@ fn wasm_equal_list(
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use clarity::vm::types::OptionalData;
+    use clarity::vm::Value;
+
+    use crate::tools::evaluate as eval;
+
+    #[test]
+    fn index_of_elem_not_in_list() {
+        assert_eq!(
+            eval("(index-of? (list 1 2 3 4 5 6 7) 9)"),
+            Some(Value::Optional(OptionalData { data: None }))
+        );
+    }
+
+    #[test]
+    fn index_of_first_elem() {
+        assert_eq!(
+            eval("(index-of? (list 1 2 3 4 5 6 7) 1)"),
+            Some(Value::Optional(OptionalData {
+                data: Some(Box::new(Value::UInt(0)))
+            }))
+        );
+    }
+
+    #[test]
+    fn index_of_elem() {
+        assert_eq!(
+            eval("(index-of? (list 1 2 3 4 5 6 7) 3)"),
+            Some(Value::Optional(OptionalData {
+                data: Some(Box::new(Value::UInt(2)))
+            }))
+        );
+    }
+
+    #[test]
+    fn index_of_last_elem() {
+        assert_eq!(
+            eval("(index-of? (list 1 2 3 4 5 6 7) 7)"),
+            Some(Value::Optional(OptionalData {
+                data: Some(Box::new(Value::UInt(6)))
+            }))
+        );
+    }
+
+    #[test]
+    fn index_of_called_by_v1_alias() {
+        assert_eq!(
+            eval("(index-of (list 1 2 3 4 5 6 7) 100)"),
+            Some(Value::Optional(OptionalData { data: None }))
+        );
+    }
+}
```

Comment:
Can you add a test where the reported type might not be the same as the actual type (again I'm not sure right now which type will be reported, but it will be useful to have). Something like:

```clarity
(define-private (find-it? (needle int) (haystack (list 10 int)))
  (index-of? haystack needle)
)
(print (find-it? 6 (list 1 2 3)))
(list 4 5 6 7)
```

I _think_ that if that incorrectly loops for the full 10 elements instead of the actual size of 3 elements, that it would find the 6 on the stack from the following list.

Reply:
Done.

Reply:
I think in that test you can remove the public function and just call the private function. But please add the extra list after it, to check if it searches past the end of the list in the stack. It's probably easier to check if we use an `asserts!` instead of a `print` like I have in my example.

Reply:
`asserts!` is not implemented yet. 

Anyway, I've added a test but ignoring it (`#[ignore]`) until the implementation is done.

Reply:
Sounds good! I did some more digging and I'm confident now (after our recent fixes) that this problem should not occur, but it will be good to have this test in case things change.